### PR TITLE
Stacking information stored in a `_stackOffsets` instance variable

### PR DESCRIPTION
**Pull request by bluong**

### CODE REVIEW

File: src/components/plots/stackedAreaPlot.ts [hunk 1]
```diff
@@ -21,6 +22,7 @@ export module Plots {
       this.classed("area-plot", true);
       this._isVertical = true;
       this.attr("fill-opacity", 1);
+      this._stackOffsets = new Utils.Map<Dataset, D3.Map<number>>();
```
Comment (aicioara) [inline src/components/plots/stackedAreaPlot.ts]: move `_stackedExtent` initialization here for consistency

File: src/components/plots/stackedAreaPlot.ts [hunk 2]
```diff
@@ -101,10 +103,10 @@ export module Plots {
 
       var yAccessor = this.y().accessor;
       var xAccessor = this.x().accessor;
-      attrToProjector["y"] = (d: any, i: number, dataset: Dataset, m: StackedPlotMetadata) =>
-        this.y().scale.scale(+yAccessor(d, i, dataset, m) + m.offsets.get(xAccessor(d, i, dataset, m)));
-      attrToProjector["y0"] = (d: any, i: number, dataset: Dataset, m: StackedPlotMetadata) =>
-        this.y().scale.scale(m.offsets.get(xAccessor(d, i, dataset, m)));
+      attrToProjector["y"] = (d: any, i: number, dataset: Dataset, m: PlotMetadata) =>
+        this.y().scale.scale(+yAccessor(d, i, dataset, m) + this._stackOffsets.get(dataset).get(xAccessor(d, i, dataset, m)));
+      attrToProjector["y0"] = (d: any, i: number, dataset: Dataset, m: PlotMetadata) =>
```
Comment (aicioara) [inline src/components/plots/stackedAreaPlot.ts]: I guess PlotMetadata will disappear completely from this signature in one of your future issues?

Comment (bluong) [inline src/components/plots/stackedAreaPlot.ts]: It is the next direct step

File: src/components/plots/stackedAreaPlot.ts [hunk 3]
```diff
@@ -129,7 +126,17 @@ export module Plots {
         var keyToPlotDatasetKey = this._key2PlotDatasetKey;
         var filter = this._filterForProperty(this._isVertical ? "y" : "x");
 
-        var extents = StackedPlotUtils.computeStackExtents(keyAccessor, valueAccessor, datasetKeys, keyToPlotDatasetKey, filter);
+        if (this._stackOffsets.keys().length === 0) {
+          var stackOffsets = StackedPlotUtils.computeStackOffsets.call(this, keyAccessor, valueAccessor, datasetKeys, keyToPlotDatasetKey);
```
Comment (aicioara) [inline src/components/plots/stackedAreaPlot.ts]: there is no need to pass `this`
```typescript
var stackOffsets = StackedPlotUtils.computeStackOffsets(keyAccessor, valueAccessor, datasetKeys, keyToPlotDatasetKey);
```

File: src/components/plots/stackedAreaPlot.ts [hunk 4]
```diff
@@ -129,7 +126,17 @@ export module Plots {
...
+
+          for (var datasetKey in stackOffsets) {
+            if (!stackOffsets.hasOwnProperty(datasetKey)) {
+              continue;
+            }
+            this._stackOffsets.set(keyToPlotDatasetKey.get(datasetKey).dataset, stackOffsets[datasetKey]);
+          }
+        }
+        var extents = StackedPlotUtils.computeStackExtents(keyAccessor, valueAccessor, this.datasets(), this._stackOffsets, filter);
```
Comment (aicioara) [inline src/components/plots/stackedAreaPlot.ts]: This is a duplication of `_updateStackExtentsAndOffsets`

File: src/components/plots/stackedBarPlot.ts [hunk 5]
```diff
@@ -16,6 +17,7 @@ export module Plots {
      */
     constructor(xScale?: Scale<X, number>, yScale?: Scale<Y, number>, isVertical = true) {
       super(xScale, yScale, isVertical);
+      this._stackOffsets = new Utils.Map<Dataset, D3.Map<number>>();
```
Comment (aicioara) [inline src/components/plots/stackedBarPlot.ts]: Same for `_stackExtent`

File: src/components/plots/stackedBarPlot.ts [hunk 6]
```diff
@@ -111,7 +108,17 @@ export module Plots {
...
+          var stackOffsets = StackedPlotUtils.computeStackOffsets.call(this, keyAccessor, valueAccessor, datasetKeys, keyToPlotDatasetKey);
+
+          for (var datasetKey in stackOffsets) {
+            if (!stackOffsets.hasOwnProperty(datasetKey)) {
+              continue;
+            }
+            this._stackOffsets.set(keyToPlotDatasetKey.get(datasetKey).dataset, stackOffsets[datasetKey]);
+          }
+        }
+        var extents = StackedPlotUtils.computeStackExtents(keyAccessor, valueAccessor, this.datasets(), this._stackOffsets, filter);
```
Comment (aicioara) [inline src/components/plots/stackedBarPlot.ts]: same as for the other file

File: test/components/plots/stackedPlotTests.ts [hunk 7]
```diff
@@ -68,17 +72,20 @@ describe("Plots", () => {
...
+      var ds0 = new Plottable.Dataset(data0);
+      var ds1 = new Plottable.Dataset(data1);
+      var ds2 = new Plottable.Dataset(data2);
+      var ds3 = new Plottable.Dataset(data3);
+      stackedPlot.addDataset(ds0);
+      stackedPlot.addDataset(ds1);
+      stackedPlot.addDataset(ds2);
+      stackedPlot.addDataset(ds3);
 
       // HACKHACK #1984: Dataset keys are being removed, so these are internal keys
```
Comment (aicioara) [inline test/components/plots/stackedPlotTests.ts]: Is this still the case?

File: test/tests.js [hunk 8]
```diff
@@ -4525,17 +4537,21 @@ describe("Plots", function () {
...
+            var ds1 = new Plottable.Dataset(data1);
+            var ds2 = new Plottable.Dataset(data2);
+            var ds3 = new Plottable.Dataset(data3);
+            var ds4 = new Plottable.Dataset(data4);
+            stackedPlot.addDataset(ds0);
+            stackedPlot.addDataset(ds1);
+            stackedPlot.addDataset(ds2);
+            stackedPlot.addDataset(ds3);
+            stackedPlot.addDataset(ds4);
             // HACKHACK #1984: Dataset keys are being removed, so these are internal keys
```
Comment (aicioara) [inline test/tests.js]: Same

**Pull request merged**